### PR TITLE
Make sure jobs triggered by airflow web are not identified as orphaned.

### LINKED ISSUE #20982: Triggered tasks from web-ui are killed by scheduler because they are orphaned task instances
### Apache Airflow version

2.2.2

### What happened

I have a dag which is scheduled every day and has depends_on_past set to true.
When a task fails for a given date, which is expected as the required data is not there. If I want to manually run the next task (for the following day) the run always fails.

The reason for this is that airflow creates next task instances for the following days and sets their state to None as they cannot be scheduled because the previous task instance is in failed state, this is correct and expected behavior. If I now manually trigger the task_run from the web-ui, the task_instance data in the database is not updated and as a consequence the queued_by_job_id is not filled in. 

Every 5 minutes the airflow scheduler queries for orphaned tasks, and since my manual run does not have queued_by_job_id filled in, it always gets killed as the scheduler thinks it is orphaned.
The scheduler shows the following logs:
`
airflow-scheduler [2022-01-20 11:15:41,784] {scheduler_job.py:1178} INFO - Reset the following 1 orphaned TaskInstances:                                                                   airflow-scheduler     <TaskInstance: testmanual.sample scheduled__2022-01-16T00:00:00+00:00 [running]> 
`

### What you expected to happen

I expect that the manual run will not be killed by the scheduler as it thinks it is orphaned and thus that my tasks can still succeed.

If this is expected behavior, It is best to show an error to the user when it tries to run the request stating: Running this task manually is not supported because... Then at least it is clear to the user, now the actual reason is really hidden and not obvious for most users I assume. 

### How to reproduce

- Create a dag that fails for the start_date and for the following dates just sleeps for 5 minutes and set the start date to yesterday.
- Enable the dag and the run for yesterday will fail (expected)
- Manually trigger the dag for today through the web-ui
- The run for today will be killed by the scheduler every time

### Operating System

debian buster

### Versions of Apache Airflow Providers

/

### Deployment

Other Docker-based deployment

### Deployment details

We run airflow on kubernetes and thus use the kubernetes_executor to schedule tasks.

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by nclaeys**
closes: #20982 

---
Updated the kubernetes_executor code to take into account manually launched tasks (through the web-ui).
This way they are not reset as orphaned tasks.

### CODE REVIEW

File: airflow/executors/kubernetes_executor.py [hunk 1]
```diff
@@ -696,8 +701,7 @@ def try_adopt_task_instances(self, tis: List[TaskInstance]) -> List[TaskInstance
             for pod in pod_list.items:
                 self.adopt_launched_task(kube_client, pod, pod_ids)
         self._adopt_completed_pods(kube_client)
-        tis_to_flush.extend(pod_ids.values())
-        return tis_to_flush
+        return list(pod_ids.values())
```
Comment (uranusjr) [inline airflow/executors/kubernetes_executor.py]: I took a look at the call sites of this, it seems that non of the actual usages strictly rely on this being a list; we should change the annotation to `Sequence[TaskInstance]` and return `pod_ids.values()` instead.

This would need to change a bunch more files so it’s OK to do this in a separate PR.

Comment (nclaeys) [reply airflow/executors/kubernetes_executor.py]: I will look at that this week and create a new PR for it.

**Comment by tanelk** [comment]
@ashb can you take a look at this bugfix. It has flown under the radar.

**Review by potiuk — APPROVED** [review]

File: tests/executors/test_kubernetes_executor.py [hunk 2]
```diff
@@ -504,6 +505,39 @@ def test_try_adopt_task_instances(self, mock_adopt_completed_pods, mock_adopt_la
         mock_adopt_completed_pods.assert_called_once()
         assert reset_tis == []  # This time our return is empty - no TIs to reset
 
+    @mock.patch('airflow.executors.kubernetes_executor.KubernetesExecutor.adopt_launched_task')
+    @mock.patch('airflow.executors.kubernetes_executor.KubernetesExecutor._adopt_completed_pods')
+    def test_try_adopt_task_instances_manual_runs(self, mock_adopt_completed_pods, mock_adopt_launched_task):
```
Comment (ashb) [inline tests/executors/test_kubernetes_executor.py]: "manual" is the wrong name here - elsewhere in the code base that commonly refers to a manual dag run, but the scheduler still handles the queuing etc.

But here your are not (I think) talking about a manual dag run, but a run action performed on a single TaskInstance, correct?

**Review by ashb — CHANGES_REQUESTED** [review]
Need to clean up naming here

File: tests/executors/test_kubernetes_executor.py (hunk 2 above)
Comment (nclaeys) [reply tests/executors/test_kubernetes_executor.py]: No indeed it is no manual dag run but indeed a task triggered by a user through a run action. I see the confusion and will rename the test.

File: airflow/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -689,9 +689,14 @@ def _change_state(self, key: TaskInstanceKey, state: Optional[str], pod_id: str,
...
 
     def try_adopt_task_instances(self, tis: List[TaskInstance]) -> List[TaskInstance]:
-        tis_to_flush = [ti for ti in tis if not ti.queued_by_job_id]
-        scheduler_job_ids = {ti.queued_by_job_id for ti in tis}
-        pod_ids = {ti.key: ti for ti in tis if ti.queued_by_job_id}
+        scheduler_job_ids = {ti.queued_by_job_id for ti in tis if ti.queued_by_job_id}
+
+        # Tasks triggered through API will have no ti.queued_by_job_id
+        # and their pod will have label 'airflow-worker=manual'
+        if any(ti for ti in tis if not ti.queued_by_job_id):
```
Comment (SamWheating) [inline airflow/executors/kubernetes_executor.py:707]: This change will fix the issue of tasks being incorrectly reset, but I think that in some cases we could still could end up with empty fields in the database, which might cause issues elsewhere or later on. 

Would it be possible instead to just set the `queued_by_job_id`  field to `manual` when the taskInstance is initially queued? We already have to manually write the `queued_dttm` for similar reasons. 

https://github.com/apache/airflow/blob/cfa95af7e83b067787d8d6596caa3bc97f4b25bd/airflow/www/views.py#L1811-L1819

Comment (nclaeys) [reply airflow/executors/kubernetes_executor.py:707]: I am not sure which issues you are talking about, which makes it difficult for me to implement it. Also if you are speaking in general about queued_by_job_id not being filled in than that is an issue for any executor, so should that logic not reside in the scheduler code itself instead of in the kubernetes_executor? At this moment I would feel uncomfortable adding that and I believe it is a better fit in a separate ticket...

**Comment by nclaeys** [comment]
@potiuk is there anything blocking the PR?

**Comment by potiuk** [comment]
> @potiuk is there anything blocking the PR?

Why asking me? If I see correctly, there are some changes requested by @ashb?

File: airflow/executors/kubernetes_executor.py [hunk 4]
```diff
@@ -689,9 +689,14 @@ def _change_state(self, key: TaskInstanceKey, state: Optional[str], pod_id: str,
...
+            scheduler_job_ids.add('manual')
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py:708]: This introduces a new race condition with multiple schedulers:

If there is a manually launched task, and two schedulers both try to adopt this task, one of them will win (as in the label will be changed) but both schedulers will end up adding the pod to their `running` list.

Comment (nclaeys) [reply airflow/executors/kubernetes_executor.py:708]: You are right, not sure how to handle that within Airflow.

Comment (tanelk) [reply airflow/executors/kubernetes_executor.py:708]: I think this check allready covers this: https://github.com/apache/airflow/blob/c95081406fa47ee737fa7937be4f13e47b2befca/airflow/executors/kubernetes_executor.py#L717-L734

It would be nice to have a UT for it

Comment (tanelk) [reply airflow/executors/kubernetes_executor.py:708]: The `pod_ids` is built from the `tis` and that is constructed with row locks.

Comment (tanelk) [reply airflow/executors/kubernetes_executor.py:697]: The `pod_ids` is built from the `tis` and that is constructed with row locks.

**Comment by potiuk** [comment]
can you please rebase @nclaeys ? there is a conflict.

**Comment by nclaeys** [comment]
> can you please rebase @nclaeys ? there is a conflict.

Resolved conflict

File: airflow/executors/kubernetes_executor.py [hunk 5]
```diff
@@ -700,9 +700,14 @@ def _change_state(self, key: TaskInstanceKey, state: Optional[str], pod_id: str,
         self.event_buffer[key] = state, None
 
     def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[TaskInstance]:
-        tis_to_flush = [ti for ti in tis if not ti.queued_by_job_id]
-        scheduler_job_ids = {ti.queued_by_job_id for ti in tis}
-        pod_ids = {ti.key: ti for ti in tis if ti.queued_by_job_id}
+        scheduler_job_ids = {ti.queued_by_job_id for ti in tis if ti.queued_by_job_id}
+
+        # Tasks triggered through API will have no ti.queued_by_job_id
+        # and their pod will have label 'airflow-worker=manual'
```
Comment (hterik) [inline airflow/executors/kubernetes_executor.py:706]: Are you sure about this? We us the REST API a lot to trigger dags and I've never seen any pod with `airflow-worker=manual`. Do you have any pointers to where the _manual_ value is created.

Comment (tanelk) [reply airflow/executors/kubernetes_executor.py:706]: I believe they mean this https://github.com/apache/airflow/blob/cb8c671eeaae44fae0f5b0135097ce4476f4141f/airflow/www/views.py#L1760-L1767

This is not related to manually triggering a dagrun, but running a single task in an existing dagrun - perhaps that caused some confusion.

**Comment by ashb** [comment]
Removing this PR form the milestone as it appears the OP has no time to finish it off.
